File: bootstrap/src/main/resources/executor/pi/src/agent-state-v2-codec.mjs
import { createHash } from "node:crypto";

export const AGENT_STATE_V2_PROTOCOL = "rd-agent-state/v2";
export const JSON_MAX_SAFE_INTEGER = 9_007_199_254_740_991;

const BUDGET_AVAILABILITY = new Set(["UNKNOWN", "AVAILABLE", "UNAVAILABLE"]);

/** RFC 8785-compatible canonical JSON for the integer-only state v2 schema. */
export function canonicalizeStateV2(value) {
  return canonical(value, "$" );
}

export function hashStateV2(value) {
  return `sha256:${createHash("sha256").update(canonicalizeStateV2(value), "utf8").digest("hex")}`;
}

export function decodeAndVerifyStateV2(json, expectedHash) {
  if (typeof json !== "string" || json.trim().length === 0) {
    throw new Error("state JSON must not be blank");
  }
  let value;
  try {
    value = JSON.parse(json);
  } catch (error) {
    throw new Error(`invalid state JSON: ${String(error.message ?? error)}`);
  }
  if (!isPlainObject(value)) {
    throw new Error("state JSON must be an object");
  }
  validateStateContract(value);
  const canonicalJson = canonicalizeStateV2(value);
  const actualHash = `sha256:${createHash("sha256").update(canonicalJson, "utf8").digest("hex")}`;
  if (actualHash !== String(expectedHash ?? "").toLowerCase()) {
    throw new Error("state hash mismatch");
  }
  if (canonicalJson !== json) {
    throw new Error("state JSON must use canonical bytes");
  }
  return value;
}

function validateStateContract(value) {
  if (value.protocol !== AGENT_STATE_V2_PROTOCOL) {
    throw new Error(`state protocol must be ${AGENT_STATE_V2_PROTOCOL}`);
  }
  if (value.budget === undefined || value.budget === null) return;
  if (!isPlainObject(value.budget)) {
    throw new Error("state budget must be an object");
  }
  if (!BUDGET_AVAILABILITY.has(value.budget.availability)) {
    throw new Error(`unsupported budget availability: ${String(value.budget.availability ?? "")}`);
  }
}

function canonical(value, path) {
  if (value === null) return "null";
  if (typeof value === "string") {
    validateUnicode(value, path);
    return JSON.stringify(value);
  }
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new Error(`${path} number exceeds JSON safe integer range or is not an integer`);
    }
    return String(value);
  }
  if (typeof value === "boolean") return String(value);
  if (Array.isArray(value)) {
    return `[${value.map((item, index) => canonical(item, `${path}[${index}]`)).join(",")}]`;
  }
  if (isPlainObject(value)) {
    return `{${Object.keys(value).sort().map((key) => {
      validateUnicode(key, `${path}.<key>`);
      return `${JSON.stringify(key)}:${canonical(value[key], `${path}.${key}`)}`;
    }).join(",")}}`;
  }
  throw new Error(`${path} contains an unsupported JSON value`);
}

function validateUnicode(value, path) {
  for (let index = 0; index < value.length; index += 1) {
    const current = value.charCodeAt(index);
    if (current >= 0xD800 && current <= 0xDBFF) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xDC00 && next <= 0xDFFF)) {
        throw new Error(`${path} contains an illegal surrogate`);
      }
      index += 1;
    } else if (current >= 0xDC00 && current <= 0xDFFF) {
      throw new Error(`${path} contains an illegal surrogate`);
    }
  }
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

File: rag/src/main/java/com/wish/rd/rag/project/agent/model/AgentStateV2Codec.java
package com.wish.rd.rag.project.agent.model;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

/** RFC 8785-compatible codec for the integer-only {@code rd-agent-state/v2} contract. */
public final class AgentStateV2Codec {

    public static final String PROTOCOL = "rd-agent-state/v2";
    public static final long MAX_SAFE_INTEGER = 9_007_199_254_740_991L;
    private static final Set<String> BUDGET_AVAILABILITY = Set.of(
            "UNKNOWN", "AVAILABLE", "UNAVAILABLE"
    );
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private AgentStateV2Codec() {
    }

    public static String canonicalize(JsonNode value) {
        StringBuilder canonical = new StringBuilder();
        appendCanonical(requireValue(value), canonical, "$" );
        return canonical.toString();
    }

    public static String hash(JsonNode value) {
        return sha256(canonicalize(value));
    }

    public static JsonNode decodeAndVerify(String json, String expectedHash) {
        if (json == null || json.isBlank()) {
            throw new IllegalArgumentException("state JSON must not be blank");
        }
        JsonNode value;
        try {
            value = OBJECT_MAPPER.readTree(json);
        } catch (JsonProcessingException exception) {
            throw new IllegalArgumentException("invalid state JSON", exception);
        }
        if (value == null || !value.isObject()) {
            throw new IllegalArgumentException("state JSON must be an object");
        }
        validateStateContract(value);
        String canonical = canonicalize(value);
        String actualHash = sha256(canonical);
        if (!actualHash.equals(expectedHash == null ? "" : expectedHash.toLowerCase())) {
            throw new IllegalArgumentException("state hash mismatch");
        }
        if (!canonical.equals(json)) {
            throw new IllegalArgumentException("state JSON must use canonical bytes");
        }
        return value;
    }

    private static void validateStateContract(JsonNode value) {
        if (!PROTOCOL.equals(value.path("protocol").asText())) {
            throw new IllegalArgumentException("state protocol must be " + PROTOCOL);
        }
        JsonNode budget = value.get("budget");
        if (budget == null || budget.isNull()) {
            return;
        }
        if (!budget.isObject()) {
            throw new IllegalArgumentException("state budget must be an object");
        }
        String availability = budget.path("availability").asText("");
        if (!BUDGET_AVAILABILITY.contains(availability)) {
            throw new IllegalArgumentException("unsupported budget availability: " + availability);
        }
    }

    private static JsonNode requireValue(JsonNode value) {
        if (value == null || value.isMissingNode()) {
            throw new IllegalArgumentException("canonical value must not be missing");
        }
        return value;
    }

    private static void appendCanonical(JsonNode node, StringBuilder target, String path) {
        if (node.isObject()) {
            target.append('{');
            TreeMap<String, JsonNode> fields = new TreeMap<>();
            Iterator<Map.Entry<String, JsonNode>> iterator = node.fields();
            while (iterator.hasNext()) {
                Map.Entry<String, JsonNode> entry = iterator.next();
                validateUnicode(entry.getKey(), path + ".<key>");
                fields.put(entry.getKey(), entry.getValue());
            }
            boolean first = true;
            for (Map.Entry<String, JsonNode> entry : fields.entrySet()) {
                if (!first) {
                    target.append(',');
                }
                appendString(entry.getKey(), target);
                target.append(':');
                appendCanonical(entry.getValue(), target, path + "." + entry.getKey());
                first = false;
            }
            target.append('}');
            return;
        }
        if (node.isArray()) {
            target.append('[');
            for (int index = 0; index < node.size(); index++) {
                if (index > 0) {
                    target.append(',');
                }
                appendCanonical(node.get(index), target, path + "[" + index + "]");
            }
            target.append(']');
            return;
        }
        if (node.isTextual()) {
            validateUnicode(node.textValue(), path);
            appendString(node.textValue(), target);
            return;
        }
        if (node.isIntegralNumber()) {
            if (!node.canConvertToLong()) {
                throw new IllegalArgumentException(path + " integer exceeds JSON safe integer range");
            }
            long value = node.longValue();
            if (value < -MAX_SAFE_INTEGER || value > MAX_SAFE_INTEGER) {
                throw new IllegalArgumentException(path + " integer exceeds JSON safe integer range");
            }
            target.append(value);
            return;
        }
        if (node.isFloatingPointNumber()) {
            throw new IllegalArgumentException(path + " must use integer JSON numbers");
        }
        if (node.isBoolean()) {
            target.append(node.booleanValue());
            return;
        }
        if (node.isNull()) {
            target.append("null");
            return;
        }
        throw new IllegalArgumentException(path + " contains an unsupported JSON value");
    }

    private static void appendString(String value, StringBuilder target) {
        try {
            target.append(OBJECT_MAPPER.writeValueAsString(value));
        } catch (JsonProcessingException exception) {
            throw new IllegalArgumentException("state string is not serializable", exception);
        }
    }

    private static void validateUnicode(String value, String path) {
        for (int index = 0; index < value.length(); index++) {
            char current = value.charAt(index);
            if (Character.isHighSurrogate(current)) {
                if (index + 1 >= value.length() || !Character.isLowSurrogate(value.charAt(index + 1))) {
                    throw new IllegalArgumentException(path + " contains an illegal surrogate");
                }
                index++;
            } else if (Character.isLowSurrogate(current)) {
                throw new IllegalArgumentException(path + " contains an illegal surrogate");
            }
        }
    }

    private static String sha256(String canonical) {
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256")
                    .digest(canonical.getBytes(StandardCharsets.UTF_8));
            return "sha256:" + HexFormat.of().formatHex(digest);
        } catch (NoSuchAlgorithmException exception) {
            throw new IllegalStateException("SHA-256 is unavailable", exception);
        }
    }
}
